Mathlib/Analysis/Calculus/FDeriv/Pow.lean

```diff
@@ -0,0 +1,235 @@
+  hasFDerivWithinAt_id _ _ |>.pow' n
+
+theorem HasFDerivAt.fun_pow' (h : HasFDerivAt f f' x) (n : ℕ) :
+    HasFDerivAt (fun x ↦ f x ^ n) (∑ i ∈ Finset.range n, f x ^ (n.pred - i) •> f' <• f x ^ i) x :=
+  match n with
+  | 0 => by simpa using hasFDerivAt_const 1 x
+  | 1 => by simpa using h
+  | n + 1 + 1 => by
+    have := h.mul' (h.fun_pow' (n + 1))
+    simp_rw [pow_succ' _ (n + 1)]
+    exact this.congr_fderiv <| aux _ _ _ _
+
+theorem HasFDerivAt.pow' (h : HasFDerivAt f f' x) (n : ℕ) :
+    HasFDerivAt (f ^ n) (∑ i ∈ Finset.range n, f x ^ (n.pred - i) •> f' <• f x ^ i) x :=
+  h.fun_pow' n
+
+theorem hasFDerivAt_pow' (n : ℕ) {x : 𝔸} :
+    HasFDerivAt (𝕜 := 𝕜) (fun x ↦ x ^ n)
+      (∑ i ∈ Finset.range n, x ^ (n.pred - i) •> ContinuousLinearMap.id 𝕜 _ <• x ^ i) x :=
+  hasFDerivAt_id _ |>.pow' n
+
+@[fun_prop]
+theorem DifferentiableWithinAt.fun_pow (hf : DifferentiableWithinAt 𝕜 f s x) (n : ℕ) :
+    DifferentiableWithinAt 𝕜 (fun x => f x ^ n) s x :=
+  let ⟨_, hf'⟩ := hf; ⟨_, hf'.pow' n⟩
+
+@[fun_prop]
+theorem DifferentiableWithinAt.pow (hf : DifferentiableWithinAt 𝕜 f s x) :
+    ∀ n : ℕ, DifferentiableWithinAt 𝕜 (f ^ n) s x :=
+  hf.fun_pow
+
+theorem differentiableWithinAt_pow (n : ℕ) {x : 𝔸} {s : Set 𝔸} :
+    DifferentiableWithinAt 𝕜 (fun x : 𝔸 => x ^ n) s x :=
+  differentiableWithinAt_id.pow _
+
+@[simp, fun_prop]
+theorem DifferentiableAt.fun_pow (hf : DifferentiableAt 𝕜 f x) (n : ℕ) :
+    DifferentiableAt 𝕜 (fun x => f x ^ n) x :=
+  differentiableWithinAt_univ.mp <| hf.differentiableWithinAt.pow n
+
+@[simp, fun_prop]
+theorem DifferentiableAt.pow (hf : DifferentiableAt 𝕜 f x) (n : ℕ) :
+    DifferentiableAt 𝕜 (f ^ n) x := hf.fun_pow n
+
+theorem differentiableAt_pow (n : ℕ) {x : 𝔸} : DifferentiableAt 𝕜 (fun x : 𝔸 => x ^ n) x :=
+  differentiableAt_id.pow _
+
+@[fun_prop]
+theorem DifferentiableOn.fun_pow (hf : DifferentiableOn 𝕜 f s) (n : ℕ) :
+    DifferentiableOn 𝕜 (fun x => f x ^ n) s := fun x h => (hf x h).pow n
+
+@[fun_prop]
+theorem DifferentiableOn.pow (hf : DifferentiableOn 𝕜 f s) (n : ℕ) :
+    DifferentiableOn 𝕜 (f ^ n) s := hf.fun_pow n
+
+theorem differentiableOn_pow (n : ℕ) {s : Set 𝔸} : DifferentiableOn 𝕜 (fun x : 𝔸 => x ^ n) s :=
+  differentiableOn_id.pow n
+
+@[simp, fun_prop]
+theorem Differentiable.fun_pow (hf : Differentiable 𝕜 f) (n : ℕ) :
+    Differentiable 𝕜 fun x => f x ^ n :=
+  fun x => (hf x).pow n
+
+@[simp, fun_prop]
+theorem Differentiable.pow (hf : Differentiable 𝕜 f) (n : ℕ) : Differentiable 𝕜 (f ^ n) :=
+  hf.fun_pow n
+
+theorem differentiable_pow (n : ℕ) : Differentiable 𝕜 fun x : 𝔸 => x ^ n :=
+  differentiable_id.pow _
+
+theorem fderiv_fun_pow' (n : ℕ) (hf : DifferentiableAt 𝕜 f x) :
+    fderiv 𝕜 (fun x ↦ f x ^ n) x
+      = (∑ i ∈ Finset.range n, f x ^ (n.pred - i) •> fderiv 𝕜 f x <• f x ^ i) :=
+  hf.hasFDerivAt.pow' n |>.fderiv
+
+theorem fderiv_pow' (n : ℕ) (hf : DifferentiableAt 𝕜 f x) :
```

Comment:
could you add `fderiv_pow_algebra` (or whatever name) for the derivative of `x -> x^n`, and same thing for the within version?